src/Microsoft.CodeQuality.Analyzers/Core/Maintainability/CodeMetricsAnalyzer.cs

```diff
@@ -0,0 +1,485 @@
+// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Analyzer.Utilities;
+using Analyzer.Utilities.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeMetrics;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeQuality.Analyzers.Maintainability.CodeMetrics
+{
+    /// <summary>
+    /// CA1501: Avoid excessive inheritance
+    /// CA1502: Avoid excessive complexity
+    /// CA1505: Avoid unmaintainable code
+    /// CA1506: Avoid excessive class coupling
+    /// </summary>
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public sealed class CodeMetricsAnalyzer : DiagnosticAnalyzer
+    {
+        internal const string CA1501RuleId = "CA1501";
+        internal const string CA1502RuleId = "CA1502";
+        internal const string CA1505RuleId = "CA1505";
+        internal const string CA1506RuleId = "CA1506";
+
+        /// <summary>
+        /// Configuration file to configure custom threshold values for supported code metrics.
+        /// For example, the below entry changes the maximum allowed inheritance depth from the default value of 5 to 10:
+        /// 
+        ///     # FORMAT:
+        ///     # 'RuleId'(Optional 'SymbolKind'): 'Threshold'
+        ///
+        ///     CA1501: 10
+        /// See CA1508 unit tests for more examples.
+        /// </summary>
+        private const string CodeMetricsConfigurationFile = "CodeMetricsConfig.txt";
```

Comment:
is this existing format? or new? if it is new, can we use new editorconfig support?

Reply:
Talked offline, the editorconfig options passed down to the compiler/analyzers is still not checked in, and will not work against 2.6.1 (which the current branch targets)